packages/contracts-bedrock/src/Safe/LivenessModule.sol

```diff
@@ -0,0 +1,183 @@
+    /// @param _prevOwner Owner that pointed to the owner to be removed in the linked list
+    /// @param _owner Owner address to be removed.
+    /// @param _threshold New threshold.
+    function _removeOwner(address _prevOwner, address _owner, uint256 _threshold) internal {
+        safe.execTransactionFromModule({
+            to: address(safe),
+            value: 0,
+            operation: Enum.Operation.Call,
+            data: abi.encodeCall(OwnerManager.removeOwner, (_prevOwner, _owner, _threshold))
+        });
+    }
+
+    /// @notice A FREI-PI invariant check enforcing requirements on number of owners and threshold.
+    function _verifyFinalState() internal view {
+        address[] memory owners = safe.getOwners();
+        uint256 numOwners = owners.length;
+        require(
+            hasMinOwners(numOwners) || (numOwners == 1 && owners[0] == fallbackOwner),
+            "LivenessModule: Safe must have the minimum number of owners, or be owned solely by the fallback owner"
+        );
+
+        // Check that the threshold is correct. This check is also correct when there is a single
+        // owner, because get75PercentThreshold(1) returns 1.
+        uint256 threshold = safe.getThreshold();
+        require(
+            threshold == get75PercentThreshold(numOwners),
+            "LivenessModule: threshold must be 75% of the number of owners"
+        );
+
+        // Check that the guard has not been changed.
+        _verifyGuard();
+    }
+
+    /// @notice Reverts if the guard address does not match the expected value.
+    function _verifyGuard() internal view {
+        require(
+            address(livenessGuard) == address(uint160(uint256(bytes32(safe.getStorageAt(GUARD_STORAGE_SLOT, 1))))),
+            "LivenessModule: guard has been changed"
+        );
+    }
+
+    /// @notice Get the previous owner in the linked list of owners
+    function _getPrevOwner(address owner, address[] memory owners) internal pure returns (address prevOwner_) {
+        for (uint256 i = 0; i < owners.length; i++) {
+            if (owners[i] != owner) continue;
+            if (i == 0) {
+                prevOwner_ = SENTINEL_OWNERS;
+                break;
+            }
+            prevOwner_ = owners[i - 1];
+        }
+    }
+
+    /// @notice For a given number of owners, return the lowest threshold which is greater than 75.
+    ///         Note: this function returns 1 for numOwners == 1.
+    function get75PercentThreshold(uint256 _numOwners) public pure returns (uint256 threshold_) {
+        threshold_ = (_numOwners * 75 + 99) / 100;
+    }
+
+    /// @notice Check if the number of owners is greater than or equal to the minimum number of owners.
+    /// @param numOwners The number of owners.
+    /// @return A boolean indicating if the number of owners is greater than or equal to the minimum number of owners.
+    function hasMinOwners(uint256 numOwners) public view returns (bool) {
```

Comment:
This function is named weirdly, `hasMinOwners` would imply no arguments and makes sense as a public function, if it takes an arg it feels like it should be internal and possibly renamed?

Reply:
`enoughOwners`, `sufficientOwners`

just some thoughts

Reply:
Agree it doesn't make sense to be public. 
For the name, I'm going to go with the very literal `isAboveMinOwners`

Reply:
c28ea28072dccdd7477813fd51aa9d480705b451